### PR TITLE
Add option to save the game immediately on death

### PR DESCRIPTION
**Pull request by ccaviness**
#### Summary
Features "Add option to save the game immediately on death"

#### Purpose of change
To make it somewhat more difficult to savescum. Currently, once the Deathcam pops up, you can quit the game and revert to your last save, avoiding death. If you're like me, this can be a very tempting way to avoid the game theme of permadeath.

#### Describe the solution
To make it harder to savescum when you mess up, this writes a save immediately on player death. This isn't quite irrevocable. If Deathcam is enabled, the save will happen right as the "Watch the last moments of your life?" dialog is displayed and there may be a way to debug your way out of that.

#### Describe alternatives you've considered
Having more willpower to not savescum, but at my age that's probably a lost cause.

#### Testing
I've tested by creating a world, loading a new character, and immediately going out to get murdered by a zombie. When the Deathcam dialog appears, I quit the game. Loading the save up again returns right to the Deathcam dialog.

### CODE REVIEW

**Comment by ccaviness** [comment]
I have approximately no experience with C++, but this does build and run in my limited testing.

**Comment by avianArsonist** [comment]
I don't see savescumming being an option as a bad thing; Permadeath doesn't have to be the only way to play imo, and making it harder to work around doesn't seem like a good idea to me

**Comment by ccaviness** [comment]
> I don't see savescumming being an option as a bad thing; Permadeath doesn't have to be the only way to play imo, and making it harder to work around doesn't seem like a good idea to me

Just to be clear (and I edited the PR title), this adds an *option*  you can enable that will save the game on death. I have no problems with savescumming - play however you want!

I'm just adding an option to make it harder to savescum that you can enable if you want.

**Comment by Night-Pryanik** [comment]
I think this should be turned on by default, because permadeath is the intended way to play the game AFAIK, and it's the savescumming should be an option, not the "normal" gameplay.

File: src/game.cpp [hunk 1]
```diff
@@ -2341,6 +2341,9 @@ bool game::is_game_over()
             return false;
         }
         Messages::deactivate();
+        if( get_option<bool> ( "AUTOSAVE_ON_DEATH" ) ) {
+            save();
```
Comment (ZhilkinSerg) [inline src/game.cpp:2345]: It should be `autosave` or `quicksave` here (so there would be some UI feedback, not just a silent save).

**Comment by Fris0uman** [comment]
Current stand on save scumming is to leave it where it is and put no work in enabling it nor disabling it.
See Kevin comment in this PR https://github.com/CleverRaven/Cataclysm-DDA/pull/39798#issuecomment-618061795

> My position on this has been "we neither support nor work to prevent savescumming" for a very long time now

**Comment by ZhilkinSerg** [comment]
> 
> 
> > I think this should be turned on by default, because permadeath is the intended way to play the game AFAIK, and it's the savescumming should be an option, not the "normal" gameplay.
> 
> Yes, after all, this game is supposed to be a roguelike.

No, after all, this game is not supposed to be a roguelike.

File: src/game.cpp (hunk 1 above)
Comment (ccaviness) [reply src/game.cpp]: Done.

**Comment by Inglonias** [comment]
What stops me from force quitting the game before I die? Cataclysm is a game where you can _usually_ see your death coming a turn or two in advance, thus rendering this fix less useful than you might think.

Also, as others have stated, this is explicitly mentioned in design docs as something that the devs do not want to be pursued.

**Comment by ccaviness** [comment]
> What stops me from force quitting the game before I die?

Nothing. Quit whenever you want!

> Cataclysm is a game where you can _usually_ see your death coming a turn or two in advance, thus rendering this fix less useful than you might think.
> Also, as others have stated, this is explicitly mentioned in design docs as something that the devs do not want to be pursued.

Not that you suggested this, but just to be clear: I didn't make this PR as some sort of anti-savescum manifesto. I have absolutely no problem with people playing this game however they want. (I added the option defaulting to `false` just for this reason.) My reason was to help enforce, just a little, the decision to play with a true permadeath in my own gameplay. I've found myself overly tempted to quit the game right at death and revert to the last save instead of restarting entirely. Again - I'm thrilled if this is how anyone else plays! I just wanted the **option** of making it a bit harder.

As you point out, this obviously doesn't stop just quitting a bit earlier. I'm not trying to "fix" all workarounds. My intent is to make it a little harder to savescum, **if** that's your gameplay style. If it's not, this option is not for you, and the default won't change a thing.

**Comment by Inglonias** [comment]
Fair enough. As a gameplay option this makes some sense to me, but I am not the person you need to convince.

**Comment by ccaviness** [comment]
As pointed out by @Fris0uman, a somewhat more complicated but similar proposal was rejected in #39798. This proposal is less complex, but perhaps this is still antithetical to the current game designer's plans.

This would be a simple patch for me to maintain locally, and I (of course) won't be bothered if this is closed.

**Comment by kevingranade** [comment]
Thanks for the thoughtfulness put into this, but no I'm sticking with the existing decision in this area.

**Pull request closed without merging**